migration-engine/core/src/state.rs

```diff
@@ -44,22 +46,31 @@ type ErasedConnectorRequest = Box<
 >;
 
 impl EngineState {
-    pub(crate) fn new(initial_datamodel: Option<String>, host: Option<Arc<dyn ConnectorHost>>) -> Self {
+    pub(crate) fn new(
+        initial_datamodel: Option<String>,
+        namespaces: Vec<String>,
+        host: Option<Arc<dyn ConnectorHost>>,
+    ) -> Self {
         EngineState {
             initial_datamodel: initial_datamodel.map(|s| psl::validate(s.into())),
+            initial_namespaces: namespaces,
             host: host.unwrap_or_else(|| Arc::new(migration_connector::EmptyHost)),
             connectors: Default::default(),
         }
     }
 
     fn namespaces(&self) -> Option<Namespaces> {
-        self.initial_datamodel
-            .as_ref()
-            .and_then(|schema| schema.configuration.datasources.first())
-            .and_then(|ds| {
-                let mut names = ds.namespaces.iter().map(|(ns, _)| ns.to_owned()).collect();
-                Namespaces::from_vec(&mut names)
-            })
+        if !self.initial_namespaces.is_empty() {
+            Namespaces::from_vec(&mut self.initial_namespaces.clone())
+        } else {
+            self.initial_datamodel
+                .as_ref()
+                .and_then(|schema| schema.configuration.datasources.first())
+                .and_then(|ds| {
+                    let mut names = ds.namespaces.iter().map(|(ns, _)| ns.to_owned()).collect();
+                    Namespaces::from_vec(&mut names)
+                })
+        }
     }
```

Comment:
Either return the namespaces with which the `EngineState` was initialized, or take them from the schema